pkg/server/cloud/gce/instances.go

```diff
@@ -0,0 +1,456 @@
+EOF
+
+    $(/usr/bin/google_instance_setup)
+    `
+	name := makeInstanceID(c.controllerID, node.Name)
+	diskType := c.getDiskTypeURL()
+	volSizeGiB := cloud.ToSaneVolumeSize(node.Spec.Resources.VolumeSize)
+	disks := c.getAttachedDiskSpec(true, int64(volSizeGiB), name, diskType, node.Spec.BootImage)
+	labels := c.getInstanceLabels(node.Name)
+	networkInterfaces := c.getInstanceNetworkSpec(node.Spec.Resources.PrivateIPOnly)
+	instanceType := c.getInstanceTypeURL(node.Spec.InstanceType)
+	spec := &compute.Instance{
+		Disks:             disks,
+		Labels:            labels,
+		MachineType:       instanceType,
+		Name:              name,
+		NetworkInterfaces: networkInterfaces,
+		Tags: &compute.Tags{
+			Items: c.bootSecurityGroupIDs,
+		},
+		Metadata: &compute.Metadata{
+			Items: []*compute.MetadataItems{
+				{
+					Key:   "user-data",
+					Value: &mds,
+				},
+				{
+					Key:   "startup-script",
+					Value: &disableIPAliases,
+				},
+			},
+		},
+	}
+	if node.Spec.Spot {
+		ar := false
+		spec.Scheduling = &compute.Scheduling{
+			AutomaticRestart:  &ar,
+			OnHostMaintenance: "TERMINATE",
+			Preemptible:       true,
+		}
+	}
+	return spec, nil
+}
+
+// this function handles the starting of both regular and spot type instances
+// it is called in the exported StartNode and StartSpotNode functions
+func (c *gceClient) startNode(node *api.Node, metadata string) (*cloud.StartNodeResult, error) {
+	klog.V(2).Infof("Starting instance for node: %v", node)
+	spec, err := c.createInstanceSpec(node, metadata)
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("Starting node with security groups: %v subnet: '%s'",
+		c.bootSecurityGroupIDs, c.subnetName)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	op, err := c.service.Instances.Insert(c.projectID, c.zone, spec).Context(ctx).Do()
+	if err != nil {
+		return nil, util.WrapError(err, "startup error")
+	}
+	// Todo: catch and convert errors to notify us of
+	// out of capacity errors or invalid machine types
+	// see pkg/server/cloud/aws/instances.StartNode()
+	if err := c.waitOnOperation(op.Name, c.getZoneOperation); err != nil {
+		return nil, err
+	}
+	startResult := &cloud.StartNodeResult{
+		InstanceID:       spec.Name,
+		AvailabilityZone: c.zone,
+	}
+	return startResult, nil
+}
+
+func (c *gceClient) StartNode(node *api.Node, metadata string) (*cloud.StartNodeResult, error) {
+	return c.startNode(node, metadata)
+}
+
+// In we dictate whether the node is a spot based on the node passed in
+// this is decided in createInstanceSpec which is called in the unexported
+// startNode function. StartSpotNode is necessary to fullfil the interface.
+func (c *gceClient) StartSpotNode(node *api.Node, metadata string) (*cloud.StartNodeResult, error) {
+	return c.startNode(node, metadata)
+}
+
+func (c *gceClient) WaitForRunning(node *api.Node) ([]api.NetworkAddress, error) {
+	for {
+		status, err := c.getInstanceStatus(node.Status.InstanceID)
```

Comment:
The call right before this is StartNode which is a write operation to the cloud.  Those writes are eventually consistent.  That means the instanceID might not _really_ exist  in GCE when we make the call to getInstanceStatus.  I'll work on figuring out the retry logic for this one.

1. What does it look like when the instance doesn't exist yet.  Can we handle that for 30s.  After 30s, if the instance doesn't exist, we should return an error.
2. We need to time out the waitForRunning operation so this eventually ends if the instance never comes up, gets deleted behind the scenes, etc.

Reply:
https://github.com/elotl/kip/commit/b4f0e5cd7bc7a7ccbb3ebc3ca0b53ecb22fd1e19